### DISCUSSION TITLE
What are the differences of ChaCha20-Poly1305 between OpenSSL 1.1.1t and OpenSSL 3.0.0

### DISCUSSION THREAD

**Discussion by WilburZjh** [openssl/openssl#21801] [Q&A]
I could not find a document to state the major difference of the implementation of ChaCha20-Poly1305 between OpenSSL 1.1.1t and OpenSSL 3.0.0. Can someone point out the difference?

**Comment by slontis** [comment]
It should be fairly similar between these 2 versions
In OpenSSL 3.0 all code to handle ciphers was rewritten to work with providers.. So instead of using the fixed method table specified in e_chacha20_poly1305.c (see EVP_chacha20_poly1305()) it now uses.
cipher_chacha20_poly1305.c & cipher_chacha20_poly1305_hw.c (The default provider uses ossl_chacha20_ossl_poly1305_functions).

There was a CVE (CVE-2019-1543) fixed in 3.0 to prevent overly long nonces being used in ChaCha20-Poly1305. (but this should have been back ported),

There are known answer test vectors in test/recipes/30-test_evp_data/evpciph_chacha.txt
